### PR TITLE
Spark Job mfe contract  to 2024-04-01 GA version

### PR DESCRIPTION
**Pull request by vaibhj**
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

> [!NOTE]
> As of January 2024 there is no PR assignee. This is expected. See https://aka.ms/azsdk/pr-arm-review.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

diagram

<details>
<summary> Click here to see the details of Step 1 </summary>

### Breaking changes review (Diagram Step 1)

If the automation determines you have breaking changes, i.e. Step 1 from the diagram applies to you,
  you must follow the [breaking changes process].  
  **IMPORTANT** This applies even if:
- The tool fails while it shouldn't, e.g. due to runtime exception, or incorrect detection of breaking changes.
- You believe there is no need for you to request breaking change approval, for any reason. 
    Such claims must be reviewed, and the process is the same.

</details>

<details> 
<summary> Click here to see the details of Step 2 </summary>

### ARM API changes review (Diagram Step 2)

- For help with Step 2 / "ARM review", see https://aka.ms/azsdk/pr-arm-review.

</details>

<details>
<summary> Click here to see the diagram footnotes</summary>

### Diagram footnotes

[1] See [ARM review queue] (for **PR merge** queues, see [2]).  
[2] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])  
The ARM reviewer on-call engineer visits the merge queue twice a day, so the approximate ETA for merges is 12 - 24 hours.

</details>

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [ ] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been generated using [OpenAPI Hub](https://aka.ms/openapihub)).
  - [x] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with PR workflow diagram Step 2 (ARM review), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://aka.ms/azsdk/pr-arm-review#about-the-arm-review-queue
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[breaking changes process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

### CODE REVIEW

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json [hunk 1]
```diff
@@ -37,17 +37,17 @@
...
-            "schema": {
-              "$ref": "#/definitions/CodeContainerResourceArmPaginatedResult"
-            }
-          },
           "default": {
             "description": "Error",
             "schema": {
               "$ref": "../../../../../common-types/resource-management/v3/types.json#/definitions/ErrorResponse"
             }
+          },
```
[the remark this answers is not in the document]
Comment (vaibhj) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:45]: This is auto generated code and 22k+ lines .. I have not made this change .. I am not sure if we can merge as generated or you want me to rearrange order.

[the remark this answers is not in the document]
Comment (vaibhj) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:45]: As per our team TSG we should not change this file manually and this file should be auto generated.  Some how autogeneration is not in right order. I already notified owning team. This should not impact in actual behavior, only order change. Please review it. Thanks, VJ

[the remark this answers is not in the document]
Comment (ms-zhenhua) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:45]: thousands of lines of order changes w/o content changes will seriously slow down the review speed. Please update the PR to only show changed content.

**Comment by vaibhj** [comment]
[lintdiff_output_2.txt](https://github.com/Azure/azure-rest-api-specs/files/14515147/lintdiff_output_2.txt)


I ran tool locally and I have result .. Can we get signoff on lintdiff using local run result . Please find attached results 
 
ran below command. 
 
(base) PS C:\code\azure-rest-api-specs> autorest --v3 --spectral --azure-validator --use=C:\Code\azure-openapi-validator\packages\azure-openapi-validator\autorest --tag=package-2024-04 C:\code\azure-rest-api-specs\specification\machinelearningservices\resource-manager\readme.md

@rkmanda @ms-zhenhua @konrad-jamrozik

**Comment by TimLovellSmith** [comment]
>           "x-nullable": true

why do you want to use x-nullable true?

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:17080 in 847ac52. [](commit_id = 847ac526150e1966a58ffffe76f7070bde0be65a, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>           "x-nullable": true

why do you want to use x-nullable true?

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:17093 in 847ac52. [](commit_id = 847ac526150e1966a58ffffe76f7070bde0be65a, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>             "x-nullable": true

why do you want to use x-nullable true?

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:17087 in 847ac52. [](commit_id = 847ac526150e1966a58ffffe76f7070bde0be65a, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>           "additionalProperties": {

Why are you using additionalProperties? Is there any schema you expect? Its not allowed unless its fully freeform and plain passthrough.

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:17085 in 847ac52. [](commit_id = 847ac526150e1966a58ffffe76f7070bde0be65a, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>           "description": "(Deprecated for Managed Online Endpoints) The maximum amount of time a request will stay in the queue in ISO 8601 format.\r\nDefaults to 500ms.\r\n(Now increase `request_timeout_ms` to account for any networking/queue delays)",

are you undeprecating it??

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:19349 in 847ac52. [](commit_id = 847ac526150e1966a58ffffe76f7070bde0be65a, deletion_comment = True)

**Comment by TimLovellSmith** [comment]
>           "x-nullable": true

why use x-nullable true? Azure guidelines are to prefer not using it where possible

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:20834 in 847ac52. [](commit_id = 847ac526150e1966a58ffffe76f7070bde0be65a, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>         "workspaceSecretName": {

Isn't it redundant with the secret URI? Are both of these writable properties? What happens if they don't match?

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:20836 in 847ac52. [](commit_id = 847ac526150e1966a58ffffe76f7070bde0be65a, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>           "x-nullable": true

why

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:21149 in 847ac52. [](commit_id = 847ac526150e1966a58ffffe76f7070bde0be65a, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>           "x-nullable": true

why

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:21158 in 847ac52. [](commit_id = 847ac526150e1966a58ffffe76f7070bde0be65a, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>           "type": "string",

How is a valid value known?

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:21153 in 847ac52. [](commit_id = 847ac526150e1966a58ffffe76f7070bde0be65a, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>           "description": "[Required] ARM resource ID of the code asset.",

use "format": "arm-id" if its an ARM resource id

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:21161 in 847ac52. [](commit_id = 847ac526150e1966a58ffffe76f7070bde0be65a, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>           "additionalProperties": {

additionalProperties is not allowed by default, please justify

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:21173 in 847ac52. [](commit_id = 847ac526150e1966a58ffffe76f7070bde0be65a, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>       "additionalProperties": false

nit, this is very verbose - additionalProperties: false can be implicit, not explicit

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:21411 in 847ac52. [](commit_id = 847ac526150e1966a58ffffe76f7070bde0be65a, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>           "pattern": "[a-zA-Z0-9_]",

ARM resource ids usually contain more characters, such as slash

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:21163 in 847ac52. [](commit_id = 847ac526150e1966a58ffffe76f7070bde0be65a, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>           "pattern": "[a-zA-Z0-9_]",

if its a file path why can't it contain '.' or '/'?

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:21351 in 847ac52. [](commit_id = 847ac526150e1966a58ffffe76f7070bde0be65a, deletion_comment = False)

**Comment by kseager** [comment]
> RM resource ids usually contain more characters, such as slash



> > ```
> >       "x-nullable": true
> > ```
> 
> why do you want to use x-nullable true?
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:17080 in [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a). [](commit_id = [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a), deletion_comment = False)

The reason for this is we have special handling in the deserialization logic of our service for how the default values get handled for types that are not nullable but not defined vs nullable. This doesn't cause auto-gen issues, is compliant, and is useful for our purposes.

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (vaibhj) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:45]: This is autogenerated code so I cannot change it manually as per policy there is work item to fix auto gen tool and avoid order change. Order change should not impact on any functionality.

**Comment by vaibhj** [comment]
> > ```
> >       "x-nullable": true
> > ```
> 
> why do you want to use x-nullable true?
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:17080 in [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a). [](commit_id = [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a), deletion_comment = False)

I am completely removing "secretsConfiguration": and "notificationSetting"

**Comment by vaibhj** [comment]
> > ```
> >         "x-nullable": true
> > ```
> 
> why do you want to use x-nullable true?
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:17087 in [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a). [](commit_id = [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a), deletion_comment = False)

I am completely removing "secretsConfiguration": and "notificationSetting"

**Comment by vaibhj** [comment]
> > ```
> >       "additionalProperties": {
> > ```
> 
> Why are you using additionalProperties? Is there any schema you expect? Its not allowed unless its fully freeform, user-defined, and plain passthrough.
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:17085 in [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a). [](commit_id = [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a), deletion_comment = False)

I am completely removing "secretsConfiguration"

**Comment by vaibhj** [comment]
> (Deprecated for Managed Online Endpoints)

This is autogenerated code and I verified its not in my solution, must be some manual step in target branch. 

@kseager @ragovada  can comment on it but it match with our source code branch

**Comment by vaibhj** [comment]
> > ```
> >       "x-nullable": true
> > ```
> 
> why use x-nullable true? Azure guidelines are to prefer not using it where possible
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:20834 in [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a). [](commit_id = [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a), deletion_comment = False)

SecretConfiguration removed from swagger

**Comment by vaibhj** [comment]
> > ```
> >     "workspaceSecretName": {
> > ```
> 
> Isn't it redundant with the secret URI? Are both of these writable properties? What happens if they don't match?
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:20836 in [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a). [](commit_id = [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a), deletion_comment = False)

SecretConfiguration removed from swagger so this

**Comment by TimLovellSmith** [comment]
> > RM resource ids usually contain more characters, such as slash
> 
> > > ```
> > >       "x-nullable": true
> > > ```
> > 
> > 
> > why do you want to use x-nullable true?
> > Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:17080 in [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a). [](commit_id = [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a), deletion_comment = False)
> 
> The reason for this is we have special handling in the deserialization logic of our service for how the default values get handled for types that are not nullable but not defined vs nullable. This doesn't cause auto-gen issues, is compliant, and is useful for our purposes.

I think the priority in this decision should be on providing a customer-friendly API.

**Comment by vaibhj** [comment]
The reason for this is we have special handling in the deserialization logic of our service for how the default values get handled for types that are not nullable but not defined vs nullable. This doesn't cause auto-gen issues, is compliant, and is useful for our purposes.

Please see kseager response in other comment

**Comment by vaibhj** [comment]
> > ```
> >       "x-nullable": true
> > ```
> 
> why
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:21158 in [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a). [](commit_id = [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a), deletion_comment = False)

The reason for this is we have special handling in the deserialization logic of our service for how the default values get handled for types that are not nullable but not defined vs nullable. This doesn't cause auto-gen issues, is compliant, and is useful for our purposes.

Please see kseager response in other comment

**Comment by vaibhj** [comment]
> arm-id

Done

**Comment by vaibhj** [comment]
> > ```
> >       "additionalProperties": {
> > ```
> 
> additionalProperties is not allowed by default, please justify
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:21173 in [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a). [](commit_id = [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a), deletion_comment = False)

This conf property used to pass Spark specific configuration. User can pass some specific properties which they want to configure 
example - "spark.driver.memory: 2g"

**Comment by raosuhas** [comment]
>           "description": "(Deprecated for Managed Online Endpoints) The maximum amount of time a request will stay in the queue in ISO 8601 format.\r\nDefaults to 500ms.\r\n(Now increase `request_timeout_ms` to account for any networking/queue delays)",

Could you please respond ?

---
In reply to: [1982008583](https://github.com/Azure/azure-rest-api-specs/pull/27930#issuecomment-1982008583) [](http://example.com/codeflow?ancestors=1982008583)

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:19349 in 847ac52. [](commit_id = 847ac526150e1966a58ffffe76f7070bde0be65a, deletion_comment = True)

**Comment by raosuhas** [comment]
>           "description": "[Required] ARM resource ID of the code asset.",

PLease fix

---
In reply to: [1982010700](https://github.com/Azure/azure-rest-api-specs/pull/27930#issuecomment-1982010700) [](http://example.com/codeflow?ancestors=1982010700)

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:21161 in 847ac52. [](commit_id = 847ac526150e1966a58ffffe76f7070bde0be65a, deletion_comment = False)

**Comment by raosuhas** [comment]
>           "additionalProperties": {

Discussed this in office hours. The additional properties are for downstream job params . Team will get back on why these cannot be defined with schema .

---
In reply to: [1982011159](https://github.com/Azure/azure-rest-api-specs/pull/27930#issuecomment-1982011159) [](http://example.com/codeflow?ancestors=1982011159)

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:21173 in 847ac52. [](commit_id = 847ac526150e1966a58ffffe76f7070bde0be65a, deletion_comment = False)

**Comment by vaibhj** [comment]
> > ```
> >       "description": "[Required] ARM resource ID of the code asset.",
> > ```
> 
> PLease fix
> 
> In reply to: [1982010700](https://github.com/Azure/azure-rest-api-specs/pull/27930#issuecomment-1982010700)
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:21161 in [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a). [](commit_id = [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a), deletion_comment = False)

This is already covered 

 "codeId": {
          "description": "[Required] arm-id of the code asset.",
          "minLength": 1,
          "pattern": "[a-zA-Z0-9_]",
          "type": "string",
          "x-ms-mutability": [
            "create",
            "read"
          ]
        },

**Comment by vaibhj** [comment]
> > ```
> >       "additionalProperties": {
> > ```
> 
> Discussed this in office hours. The additional properties are for downstream job params . Team will get back on why these cannot be defined with schema .
> 
> In reply to: [1982011159](https://github.com/Azure/azure-rest-api-specs/pull/27930#issuecomment-1982011159)
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:21173 in [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a). [](commit_id = [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a), deletion_comment = False)



> > ```
> >       "additionalProperties": {
> > ```
> 
> Discussed this in office hours. The additional properties are for downstream job params . Team will get back on why these cannot be defined with schema .
> 
> In reply to: [1982011159](https://github.com/Azure/azure-rest-api-specs/pull/27930#issuecomment-1982011159)
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:21173 in [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a). [](commit_id = [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a), deletion_comment = False)

As we discussed on call .. User pass string key value pair in dictionary and we do check all item in dictionary before passing it to down stream.

**Comment by vaibhj** [comment]
> > ```
> >       "description": "(Deprecated for Managed Online Endpoints) The maximum amount of time a request will stay in the queue in ISO 8601 format.\r\nDefaults to 500ms.\r\n(Now increase `request_timeout_ms` to account for any networking/queue delays)",
> > ```
> 
> Could you please respond ?
> 
> In reply to: [1982008583](https://github.com/Azure/azure-rest-api-specs/pull/27930#issuecomment-1982008583)
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:19349 in [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a). [](commit_id = [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a), deletion_comment = True)

Requested concerning team .. They will fix it.

**Comment by raosuhas** [comment]
> > > ```
> > >       "description": "[Required] ARM resource ID of the code asset.",
> > > ```
> > 
> > 
> > PLease fix
> > In reply to: [1982010700](https://github.com/Azure/azure-rest-api-specs/pull/27930#issuecomment-1982010700)
> > Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:21161 in [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a). [](commit_id = [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a), deletion_comment = False)
> 
> This is already covered
> 
> "codeId": { "description": "[Required] arm-id of the code asset.", "minLength": 1, "pattern": "[a-zA-Z0-9_]", "type": "string", "x-ms-mutability": [ "create", "read" ] },

What do mean this is not fixed ? I dont see the format tag ?

**Comment by raosuhas** [comment]
> > > ```
> > >       "additionalProperties": {
> > > ```
> > 
> > 
> > Discussed this in office hours. The additional properties are for downstream job params . Team will get back on why these cannot be defined with schema .
> > In reply to: [1982011159](https://github.com/Azure/azure-rest-api-specs/pull/27930#issuecomment-1982011159)
> > Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:21173 in [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a). [](commit_id = [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a), deletion_comment = False)
> 
> > > ```
> > >       "additionalProperties": {
> > > ```
> > 
> > 
> > Discussed this in office hours. The additional properties are for downstream job params . Team will get back on why these cannot be defined with schema .
> > In reply to: [1982011159](https://github.com/Azure/azure-rest-api-specs/pull/27930#issuecomment-1982011159)
> > Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:21173 in [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a). [](commit_id = [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a), deletion_comment = False)
> 
> As we discussed on call .. User pass string key value pair in dictionary and we do check all item in dictionary before passing it to down stream.

You were supposed to get back with what checks are you doing in the dictionary and why this cant be enforced in schema. Please document it here for future reviewers reference

**Comment by raosuhas** [comment]
> > > ```
> > >       "description": "(Deprecated for Managed Online Endpoints) The maximum amount of time a request will stay in the queue in ISO 8601 format.\r\nDefaults to 500ms.\r\n(Now increase `request_timeout_ms` to account for any networking/queue delays)",
> > > ```
> > 
> > 
> > Could you please respond ?
> > In reply to: [1982008583](https://github.com/Azure/azure-rest-api-specs/pull/27930#issuecomment-1982008583)
> > Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:19349 in [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a). [](commit_id = [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a), deletion_comment = True)
> 
> Requested concerning team .. They will fix it.

PLease restart ARM review once this is fixed

**Comment by vaibhj** [comment]
> "[a-zA-Z0-9_]",



> > > > ```
> > > >       "additionalProperties": {
> > > > ```
> > > 
> > > 
> > > Discussed this in office hours. The additional properties are for downstream job params . Team will get back on why these cannot be defined with schema .
> > > In reply to: [1982011159](https://github.com/Azure/azure-rest-api-specs/pull/27930#issuecomment-1982011159)
> > > Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:21173 in [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a). [](commit_id = [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a), deletion_comment = False)
> > 
> > 
> > > > ```
> > > >       "additionalProperties": {
> > > > ```
> > > 
> > > 
> > > Discussed this in office hours. The additional properties are for downstream job params . Team will get back on why these cannot be defined with schema .
> > > In reply to: [1982011159](https://github.com/Azure/azure-rest-api-specs/pull/27930#issuecomment-1982011159)
> > > Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:21173 in [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a). [](commit_id = [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a), deletion_comment = False)
> > 
> > 
> > As we discussed on call .. User pass string key value pair in dictionary and we do check all item in dictionary before passing it to down stream.
> 
> You were supposed to get back with what checks are you doing in the dictionary and why this cant be enforced in schema. Please document it here for future reviewers reference

It is spark contract. The Spark configuration is designed as a property bag with key value pair in the open-source. 
[Configuration - Spark 3.5.1 Documentation (apache.org)](https://spark.apache.org/docs/latest/configuration.html#available-properties)

Spark properties are not strongly typed for several reasons:
User can use any environment property in job and spark have different runtime and each runtime may have some new property .. With each version change we can not update contract if we will keep it schematized. 
Flexibility: Spark is designed to handle diverse workloads, including batch processing, streaming, machine learning, and graph processing. A rigid type system would limit its flexibility.
Dynamic Configuration: Spark allows dynamic configuration changes during runtime. Strong typing would make this more cumbersome.
Backward Compatibility: Some configuration keys have been renamed over time. Accepting both old and new key names ensures backward compatibility.
User-Defined Properties: Users can define custom properties, and enforcing strong typing for these would be challenging.
Trade-offs: While strong typing provides safety, it can also lead to verbosity and hinder expressiveness.

Validation: 

Some properties are must for starting any spark job we validate all those properties for right value including type and null check. Some properties which user want to promote to environment and if they are not required so we trust on user intention and we pass that as is to spark environment.

**Comment by vaibhj** [comment]
> > ```
> >       "description": "(Deprecated for Managed Online Endpoints) The maximum amount of time a request will stay in the queue in ISO 8601 format.\r\nDefaults to 500ms.\r\n(Now increase `request_timeout_ms` to account for any networking/queue delays)",
> > ```
> 
> Could you please respond ?
> 
> In reply to: [1982008583](https://github.com/Azure/azure-rest-api-specs/pull/27930#issuecomment-1982008583)
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:19349 in [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a). [](commit_id = [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a), deletion_comment = True)

Fixed

**Comment by vaibhj** [comment]
> > > > ```
> > > >       "description": "[Required] ARM resource ID of the code asset.",
> > > > ```
> > > 
> > > 
> > > PLease fix
> > > In reply to: [1982010700](https://github.com/Azure/azure-rest-api-specs/pull/27930#issuecomment-1982010700)
> > > Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:21161 in [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a). [](commit_id = [847ac52](https://github.com/Azure/azure-rest-api-specs/commit/847ac526150e1966a58ffffe76f7070bde0be65a), deletion_comment = False)
> > 
> > 
> > This is already covered
> > "codeId": { "description": "[Required] arm-id of the code asset.", "minLength": 1, "pattern": "[a-zA-Z0-9_]", "type": "string", "x-ms-mutability": [ "create", "read" ] },
> 
> What do mean this is not fixed ? I dont see the format tag ?

Now its covered

**Comment by raosuhas** [comment]
Just adding the tag to remove it from the review queue. Please abandon this PR once your other PR is signed off

**Pull request closed without merging**